File: src/main/java/me/han/muffin/client/event/events/network/PacketEvent.java
package me.han.muffin.client.event.events.network;

import me.han.muffin.client.event.EventCancellable;
import net.minecraft.network.Packet;

public class PacketEvent extends EventCancellable {
    public final Packet<?> packet;

    private PacketEvent(Packet<?> packet) {
        this.packet = packet;
    }

    public static class Send extends PacketEvent {
        private final EventStage stage;
        public Send(Packet<?> packet, EventStage stage) {
            super(packet);
            this.stage = stage;
        }

        @Override
        public EventStage getStage() {
            return stage;
        }

    }

    public static class Receive extends PacketEvent {
        private final EventStage stage;

        public Receive(Packet<?> packet, EventStage stage) {
            super(packet);
            this.stage = stage;
        }

        @Override
        public EventStage getStage() {
            return stage;
        }

    }

    public Packet<?> getPacket() {
        return packet;
    }

}

File: src/main/kotlin/com/defrag/client/module/modules/events/PacketEvent.kt
package com.defrag.client.module.modules.events

import com.defrag.client.module.modules.EventStage
import net.minecraft.network.Packet
import net.minecraftforge.fml.common.eventhandler.Cancelable

open class PacketEvent(stage: Int, val packet: Packet<*>) : EventStage(stage) {
    @JvmName("getPacket1") fun <T : Packet<*>?> getPacket(): Packet<*> {
        return packet
    }

    @Cancelable
    class Send(stage: Int, packet: Packet<*>) : PacketEvent(stage, packet)

    @Cancelable
    class Receive(stage: Int, packet: Packet<*>) : PacketEvent(stage, packet)
}